Plugins/Public/kill_tracker/Main.cpp

```diff
@@ -0,0 +1,419 @@
+		clientId = HkGetClientIdFromCharname(targetCharName);
+		if (!clientId)
+		{
+			PrintUserCmdText(client, L"Player not found");
+			return;
+		}
+	}
+	else
+	{
+		clientId = client;
+	}
+	int kills;
+	pub::Player::GetNumKills(clientId, kills);
+	PrintUserCmdText(client, L"PvP kills: %u", kills);
+}
+
+/** @ingroup KillTracker
+ * @brief Hook on ShipDestroyed. Increments the number of kills of a player if there is one.
+ */
+void __stdcall ShipDestroyed(DamageList* dmg, DWORD* ecx, uint kill)
+{
+	returncode = DEFAULT_RETURNCODE;
+	if (kill == 1)
+	{
+		const CShip* cShip = reinterpret_cast<CShip*>(ecx[4]);
+
+		if (uint client = cShip->GetOwnerPlayer())
+		{
+			uint lastInflictorId = dmg->get_cause() == 0 ? ClientInfo[client].dmgLast.get_inflictor_id() : dmg->get_inflictor_id();
+			uint killerId = HkGetClientIDByShip(lastInflictorId);
+
+			if (killerId && killerId != client)
+			{
+				int kills;
+				pub::Player::GetNumKills(killerId, kills);
+				kills++;
+				pub::Player::SetNumKills(killerId, kills);
+			}
+		}
+	}
+}
+
+void __stdcall AddDamageEntry(DamageList* damageList, ushort subObjId, float& newHitPoints, enum DamageEntry::SubObjFate fate)
+{
+	returncode = DEFAULT_RETURNCODE;
+	if (iDmgTo && subObjId == 1 && g_LastHitPts > newHitPoints) //ignore negative hp events such as repair ship
+	{
+		const auto& inflictor = damageList->iInflictorPlayerID;
+		if (inflictor && iDmgTo && inflictor != iDmgTo)
+		{
+			damageArray[inflictor][iDmgTo] += g_LastHitPts - newHitPoints;
+		}
+	}
+}
+
+void clearDamageTaken(uint victim)
+{
+	damageArray[victim].fill(0.0f);
+}
+
+void clearDamageDone(uint inflictor)
+{
+	for (int i = 1; i <= MAX_CLIENT_ID; i++)
+		damageArray[i][inflictor] = 0.0f;
+}
+
+void ProcessNonPvPDeath(const wstring& message, uint system)
+{
+	wstring deathMessage = L"<TRA data=\"0x0000CC01"
+		L"\" mask=\"-1\"/><TEXT>" + XMLText(message) + L"</TEXT>";
+
+	PlayerData* pd = nullptr;
+	while (pd = Players.traverse_active(pd))
+	{
+		if (pd->iSystemID == system)
+		{
+			HkFMsg(pd->iOnlineID, deathMessage);
+		}
+	}
+}
+
+wstring RandomizeDeathTemplate(uint iClientID)
+{
+	uint shipType = Archetype::GetShip(Players[iClientID].iShipArchetype)->iArchType;
+	const auto& deathMsgList = shipClassToDeathMsgListMap[shipType];
+	if (deathMsgList.empty())
+	{
+		return defaultDeathDamageTemplate;
+	}
+	else
+	{
+		return deathMsgList.at(rand() % deathMsgList.size());
+	}
+}
+
+void __stdcall SendDeathMessage(const wstring& message, uint system, uint clientVictim, uint clientKiller)
+{
+	returncode = DEFAULT_RETURNCODE;
+	if (!clientVictim)
+	{
+		return;
+	}
+
+	returncode = NOFUNCTIONCALL;
+
+	map<float, uint> damageToInflictorMap; // damage is the key instead of value because keys are sorted, used to render top contributors in order
+	set<CPlayerGroup*> killerGroups;
+
+	float totalDamageTaken = 0.0f;
+	PlayerData* pd = nullptr;
+	while (pd = Players.traverse_active(pd))
+	{
+		float damageDone = damageArray[pd->iOnlineID][clientVictim];
```

Comment:
Can you assert that damageArray always contains *both* of these values? If you get it wrong that's a crash.

Reply:
It's a square array with compile-time guarantee of being of size [MAX_CLIENT_ID + 1] aka [250][250]